### DISCUSSION TITLE
add new cross sections to .xns11 file by copying existing

### DISCUSSION THREAD

**Discussion by Snowthe** [DHI/mikeio1d#150] [Q&A]
I have an existing cross section file, to which I want to add some additional cross sections to.
I load the file via 
```
from mikeio1d.xns11 import Xns11
xns = Xns11(file_path.xns11)
```

I tried:

1. To create new cross sections from scratch via 
```
from mikeio1d.cross_sections import CrossSection
xs_new = CrossSection.from_xz(x, z, location_id, chainage, topo_id)
```
This, however, does not allow me to set other things such as zmin/zmax which we also use in our setups.
2. To create new cross sections as a copy of existing ones (as in my current application, I only need to modify the raw data of existing ones) via
```
import copy
xs_sel = xns.xsections.sel(...)
xs_new = copy.copy(xs_sel) #copy.deepcopy() throws error: TypeError: cannot pickle 'CrossSectionLocated' object
xs_new.raw = new_raw_df
```
This, however, updates also the old cross section (as deepcopy would be required?)

Am I overlooking something completely? Does any of you have a suggestion on how this could be implemented?

I even tried to add a "setter" for the topo_id property, to allow the following:
`xs_new.topo_id = 'VP3L_wType'`
(which so far, results in AttributeError: property 'topo_id' of 'CrossSection' object has no setter )
by adding 
```
    @topo_id.setter
    def topo_id(self, value: str | str):
        self._m1d_cross_section.TopoID = value
```
to cross_section.py. But that still doesn't solve the problem with copying...

**Comment by ryan-kipawa** [comment]
Hi @Snowthe,

Thanks for the question. I think I can give a workaround, but I'm also interested in how we could make this workflow easier. 

**_For the workaround:_**

I don't believe deep copies are supported for cross sections in mikeio1d yet. However, you can make a deep copy of its underlying C# object, and then create a new CrossSection from that:

```
from mikeio1d.cross_sections import CrossSection

# assumes the cross section you want to copy is called 'xs'
xs2 = CrossSection(xs.m1d_cross_section.DeepClone())
```

From here you could update the topo IDs and such via:
```
xs2.m1d_cross_section.TopoID = "new_topo"
xs2.m1d_cross_section.Location.Chainage = 1000.0
xs2.m1d_cross_section.Location.ID = "location_id"
```

_**For a better workflow**_

I wonder if you have any suggestions? Some thoughts I have are:

- Is copying existing cross sections to modify them a common workflow we should design for?
- How are ZMin and Zmax used?

I partly wonder if copying cross sections and using Zmin/Zmax was an efficient way for GUIs. Now that we can use Python, I was thinking a better approach would be to allow cross sections to easily be built from the ground up. But curious to hear your thoughts on that general approach as well 😄

**Comment by Snowthe** [comment]
Thanks a lot for the follow up. Based on your suggestion, I tried to execute the following code:
```
import mikeio1d
from mikeio1d.xns11 import Xns11
from mikeio1d.cross_sections import CrossSection

fp = <...>
xns = Xns11(fp)
# get specific cross section
xs = xns.xsections.sel(topo_id='VP3L_Type', location_id='Fladet', chainage=50)
xs2 = CrossSection(xs.m1d_cross_section.DeepClone())
```
However, this fails with the error message AttributeError: 'CrossSection' object has no attribute 'm1d_cross_section'
I run mikeio1d 0.9.1 with python 3.11
(I also noticed in the two notebook examples on viewing/editing xns11 files, that I cannot execute much of the code due to similar not implemented errors; it seems they are based on a different/older version of mikeio1d?)

Else:
- Yes, copying cross sections to modify them can be considered a common workflow
- Zmin and Zmax are for us used to define the datum (river bank level or bottom). I.e. all cross sections are defined relative to their datum, which e.g. allows to place "type cross sections" in standardized form always at topography, which is defined via the zmin/zmax (=datum in the GUI?). Whereas the raw data always remains identical. We sometimes use different topography files for different SHE model versions (=resolutions), and using the zmin/datum allows for an easy fit of cross sections to new topography datasets

**Reply by ryan-kipawa** [reply]
Great, thanks for the clarifications.

Regarding the errors you're getting: this is actually a case where the solution I pitched and notebooks are a bit ahead of 0.9.1. Quite shortly, 0.10.0 will be released and it will work fine.

**Reply by Snowthe** [reply]
Great, looking forward to the new release then! :)
